File: src/core/time.ts
/**
 * Source map in this file:
 * - stepTimestamp: adapted from rajiko/modules/timeshift.js (seek)
 * - parseTimestamp/formatTimestamp: new CLI helpers
 */
/**
 * Source: New helper for timestamp parsing in CLI.
 */
export function parseTimestamp(ts: string): Date {
  const m = /^(\d{4})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})$/.exec(ts);
  if (!m) {
    throw new Error(`Invalid timestamp: ${ts}`);
  }
  const [, y, mo, d, h, mi, s] = m;
  return new Date(
    Number(y),
    Number(mo) - 1,
    Number(d),
    Number(h),
    Number(mi),
    Number(s),
    0,
  );
}

/**
 * Source: New helper for timestamp formatting in CLI.
 */
export function formatTimestamp(date: Date): string {
  const y = date.getFullYear().toString().padStart(4, "0");
  const m = (date.getMonth() + 1).toString().padStart(2, "0");
  const d = date.getDate().toString().padStart(2, "0");
  const h = date.getHours().toString().padStart(2, "0");
  const mi = date.getMinutes().toString().padStart(2, "0");
  const s = date.getSeconds().toString().padStart(2, "0");
  return `${y}${m}${d}${h}${mi}${s}`;
}

/**
 * Source: rajiko/modules/timeshift.js -> seek (adapted)
 */
export function stepTimestamp(ts: string, seconds: number): string {
  const date = parseTimestamp(ts);
  date.setSeconds(date.getSeconds() + seconds);
  return formatTimestamp(date);
}

File: internal/util/time.go
package util

import (
	"fmt"
	"time"
)

// Source map in this file:
// - StepTimestamp behavior is adapted from rajiko/modules/timeshift.js seek stepping.
// - Parse/Format helpers are CLI utilities.
const tsLayout = "20060102150405"

// ParseTimestamp parses a local-time timestamp in "YYYYMMDDHHMMSS" format.
//
// Input must be exactly 14 digits and is interpreted in time.Local.
func ParseTimestamp(ts string) (time.Time, error) {
	if len(ts) != 14 {
		return time.Time{}, fmt.Errorf("invalid timestamp: %s", ts)
	}
	t, err := time.ParseInLocation(tsLayout, ts, time.Local)
	if err != nil {
		return time.Time{}, fmt.Errorf("invalid timestamp: %s", ts)
	}
	return t, nil
}

// FormatTimestamp formats t as "YYYYMMDDHHMMSS".
func FormatTimestamp(t time.Time) string {
	return t.Format(tsLayout)
}

// StepTimestamp shifts a "YYYYMMDDHHMMSS" timestamp by seconds and returns the
// same layout.
func StepTimestamp(ts string, seconds int) (string, error) {
	t, err := ParseTimestamp(ts)
	if err != nil {
		return "", err
	}
	return FormatTimestamp(t.Add(time.Duration(seconds) * time.Second)), nil
}
